hop/io.py

```diff
@@ -3,44 +3,218 @@
+                opts["start_at"] = start_at if start_at else self.start_at
+            if persist is not None or self.persist is not None:
+                opts["read_forever"] = persist if persist is not None else self.persist
+            return _open_consumer(
+                group_id,
+                broker_addresses,
+                topics,
+                auth=auth,
+                metadata=metadata,
+                **opts,
+            )
+        else:
+            raise ValueError("mode must be either 'w' or 'r'")
+
+
+class Deserializer(Enum):
+    CIRCULAR = models.GCNCircular
+    VOEVENT = models.VOEvent
+    BLOB = models.MessageBlob
+
+    @classmethod
+    def deserialize(cls, message):
+        """Deserialize a stream message and instantiate a model.
+
+        Args:
+            message: A serialized message.
+
+        Returns:
+            A data container corresponding to the format in the
+            serialized message.
+
+        Raises:
+            ValueError: If the message is incorrectly formatted or
+                if the message format is not recognized.
 
         """
-        opts = {**self._options, **kwargs}
-        return streaming.open(*args, **opts)
+        try:
+            format = message["format"].upper()
+            content = message["content"]
+        except (KeyError, TypeError):
+            raise ValueError("Message is incorrectly formatted")
+        else:
+            if format in cls.__members__:
+                return cls[format.upper()].value(**content)
+            else:
+                raise ValueError(f"Message format {format} not recognized")
+
+    def load(self, input_):
+        return self.value.load(input_)
+
+    def load_file(self, input_file):
+        return self.value.load_file(input_file)
+
+
+def _generate_group_id(n):
+    """Generate a random Kafka group ID.
+
+    Args:
+        n: Length of randomly generated string.
+
+    Returns:
+        The generated group ID.
+
+    """
+    alphanum = string.ascii_uppercase + string.digits
+    rand_str = ''.join(random.SystemRandom().choice(alphanum) for _ in range(n))
+    return '-'.join((getpass.getuser(), rand_str))
+
+
+_Metadata = namedtuple("Metadata", "topic partition offset timestamp key")
+
+
+class _Consumer(consumer.Consumer):
+    def stream(self, metadata=False, **kwargs):
+        for message in super().stream(**kwargs):
+            yield self.unpack(message)
+
+    @staticmethod
+    def unpack(message, metadata=False):
+        payload = json.loads(message.value().decode("utf-8"))
+        payload = Deserializer.deserialize(payload)
+        if metadata:
+            return (
+                payload,
+                _Metadata(
+                    message.topic(),
+                    message.partition(),
+                    message.offset(),
+                    message.timestamp()[1],
+                    message.key(),
+                )
+            )
+        else:
+            return payload
```

Comment:
How does a user of this API commit their message offset read position?

Reply:
In short, they don't. That wasn't supported before either, but I don't see a reason why we couldn't support that now. In principle, that shouldn't be hard to do by overriding `adc.Consumer.mark_done()` to take in a `Metadata` object instead of `confluent_kafka.Message()` since it can build a `TopicPartition` from that and store the offset that way.